File: 7.API-DESIGN-RestAPI/code/typescript/post_the_only_non_idempotent_one_2.ts
import { Request, Response } from 'express';

export async function createPayment(req: Request, res: Response) {
    const key = req.get('Idempotency-Key'); // client-generated UUID
    if (!key) {
        return res.status(400).json({ error: "Idempotency-Key header is required" });
    }

    // already processed this exact request? return the SAME result, don't re-charge
    const prior = await idemStore.get(key);
    if (prior) {
        return res.status(prior.status).json(prior.body);
    }

    const { amount } = req.body;

    const payment = await charge(amount); // the real, non-idempotent side effect
    await idemStore.save(key, 201, payment); // remember it, keyed by the idempotency key
    
    return res.status(201).json(payment);
}

// Mock store/functions for compilation
const idemStore = {
    get: async (key: string): Promise<any> => null,
    save: async (key: string, status: number, data: any): Promise<void> => {}
};
const charge = async (amount: number): Promise<any> => ({ amount, id: "pay_123" });

File: 7.API-DESIGN-RestAPI/code/rust/post_the_only_non_idempotent_one_2.rs
use actix_web::{web, HttpRequest, HttpResponse, Responder};
use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
pub struct PaymentRequest {
    pub amount: i32,
}

#[derive(Serialize, Clone)]
pub struct PaymentResponse {
    pub id: String,
    pub amount: i32,
}

#[derive(Clone)]
pub struct PriorResponse {
    pub status: u16,
    pub body: serde_json::Value,
}

pub async fn create_payment(
    req: HttpRequest,
    body: web::Json<PaymentRequest>,
) -> impl Responder {
    // client-generated UUID
    let key = match req.headers().get("Idempotency-Key") {
        Some(k) => k.to_str().unwrap_or(""),
        None => return HttpResponse::BadRequest().json(serde_json::json!({"error": "Idempotency-Key header required"})),
    };

    // already processed this exact request? return the SAME result, don't re-charge
    if let Some(prior) = idem_store_get(key).await {
        return HttpResponse::build(actix_web::http::StatusCode::from_u16(prior.status).unwrap())
            .json(prior.body);
    }

    let payment = charge(body.amount).await; // the real, non-idempotent side effect
    
    idem_store_save(key, 201, serde_json::to_value(&payment).unwrap()).await; // remember it, keyed by the idempotency key
    
    HttpResponse::Created().json(payment) // 201 Created
}

// Mocks
async fn idem_store_get(_key: &str) -> Option<PriorResponse> { None }
async fn idem_store_save(_key: &str, _status: u16, _body: serde_json::Value) {}
async fn charge(amount: i32) -> PaymentResponse {
    PaymentResponse { id: "pay_123".to_string(), amount }
}
